Daily Arbitrum Stablecoin Balances

Synthquest opened the pull request:
## Thank you for contributing to Spellbook 🪄

### Update!
Please build spells in the proper [subproject](../dbt_subprojects/) directory. For more information, please see the main [readme](../README.md), which also links to a GH discussion with the option to ask questions.

### Contribution type
Please check the type of contribution this pull request is for:

- [x] New spell(s)
- [ ] Adding to existing spell lineage
- [ ] Bug fix

**Note:** You can safely discard any section below which doesn't apply based on selection above

---

### For new spell(s)
If you are building new spell(s), please provide the following information:

- **Spell name(s):** stablecoins_arbitrum.balances
- **Description:** materialized table of daily balances of stablecoins
- **Who are the new spell(s) for?** Internal team
- **How will the new spell(s) be used downstream?** Using solely the daily_balances table hits runtime errors & recreating the logic for daily views of stablecoin balances among wallets hits query complexity/runtime errors.

This query is for anyone interested in network views of stablecoin balances
- **Implementation details:** The stablecoins are defined in the tokens_{network}.erc20_stablecoins tables
PR here: https://github.com/duneanalytics/spellbook/pull/6723#event-14312322687

These tokens will utilize the balances_incremental_subset_daily macro to generate materialized daily tables of tokens listed ^ 

Following the same approach as here:
https://github.com/duneanalytics/spellbook/pull/6608

- **Test instructions:** unique address/balance/day
- **Related issue(s):** [Link to related issues, if any]

---

### For adding to existing spell lineage
If you are adding to an existing spell lineage, please provide the following information:

- **Description:** [Description of the changes made]

---

### For bug fixes
If you are fixing a bug, please provide the following information:

- **Description:** [Description of the bug fix]
- **Steps to reproduce:** [How to reproduce the bug]
- **Implementation details:** [Information on how the bug was fixed]
- **Test instructions:** [How to test the fix]
- **Related issue(s):** [Link to related issues, if any]

---

### Additional information
Please provide any additional information that might help us review your pull request:

- [Any additional information]

---

Thank you for your contribution!

dune-eng commented:
Workflow run id 10929899636 approved.

dune-eng commented:
Workflow run id 10929899630 approved.

Synthquest commented:
@jeff-dude 

I'm planning to test one chain and run through the rest of them in a separate PR if it looks good

jeff-dude commented:
gm @Synthquest 
thank you for using the macro here 🙏 
do you think we can move this code to the daily subproject, as the data is at daily level? it is also heavy code to compute, especially as our tokens prod env is rather small compared to daily.

dune-eng commented:
Workflow run id 10944670007 approved.

dune-eng commented:
Workflow run id 10946158999 approved.

Synthquest commented:
@jeff-dude 
I moved to the daily folder and seem to have a longer runtime + a "remove" error popping up.

Chatgpt seems to think it timed out or his some resource constraint 😅. On the previous run it was able to initialize the model but not able to complete the incremental run

jeff-dude commented:
> @jeff-dude I moved to the daily folder and seem to have a longer runtime + a "remove" error popping up.
> 
> Chatgpt seems to think it timed out or hit some resource constraint 😅. On the previous run it was able to initialize the model but not able to complete the incremental run

it could've been bad timing with other PRs running code at the same time, on the same cluster. you can always try to monitor that via the actions tab. let's see how this latest run goes.

dune-eng commented:
Workflow run id 10950558230 approved.

dune-eng commented:
Workflow run id 10950558467 approved.

dune-eng commented:
Workflow run id 10952278872 approved.

dune-eng commented:
Workflow run id 10952278792 approved.

Synthquest commented:
@jeff-dude 
I limited it down to just USDC for testing - we may just have to limit the amount of stables we try and track here

jeff-dude commented:
> @jeff-dude I limited it down to just USDC for testing - we may just have to limit the amount of stables we try and track here

i don't think we need to limit ourselves on the stablecoins, it's a pretty small list. i haven't seen a commit here yet where you push a hardcoded date into the macro -- can you try that?
i.e. `block_time >= date '2024-08-01'` outside of the incrmental logic, so history is much shorter

dune-eng commented:
Workflow run id 10961562717 approved.

dune-eng commented:
Workflow run id 10961562733 approved.

Synthquest commented:
@jeff-dude 

I added in a more recent start date. For context this spell was able to fully compile in the token folder and was able to initialize in the daily folder on the first run but failed at the incremental test.

shortening the time period seems to have worked as well. Not 100% sure this is what you meant though

jeff-dude commented:
> shortening the time period seems to have worked as well. Not 100% sure this is what you meant though

yeah, that commit looks like what i'm picturing. how do the results look? have you queried the CI output table to verify? if so, we can remove the short timeframe and prep for merge

Synthquest commented:
Awesome, I quickly did a check on:
1) if top wallet balances lined up 
2) a few random balances for some of the other stables
3) outlier on max() values on balances
4) negative balances

For future guidance when we add more chains should we follow a similar approach (1 PR per chain with shortened timeframe for review)?

dune-eng commented:
Workflow run id 10965416955 approved.

dune-eng commented:
Workflow run id 10965417046 approved.

Synthquest commented:
@jeff-dude 
All checks passed 

re-
> For future guidance when we add more chains should we follow a similar approach (1 PR per chain with shortened timeframe for review)?

jeff-dude commented:
> For future guidance when we add more chains should we follow a similar approach (1 PR per chain with shortened timeframe for review)?

you can submit multiple chains per PR, if using same macro. you would more than likely have to restrict timeframe to see it run, yes

jeff-dude commented on dbt_subprojects/daily_spellbook/models/_sector/stablecoins/arbitrum/stablecoins_arbitrum_sources.yml:
rather than introduce a new source file, you should be able to list in the existing tokens subproject file:
https://github.com/duneanalytics/spellbook/blob/main/sources/_subprojects_outputs/tokens/_sources.yml

the goal with this directory of sources is to put spells that are used across dbt projects. we try to put sources in the directory path of the project which they are built in.

Synthquest replied:
Updated in this commit:
https://github.com/duneanalytics/spellbook/pull/6757/commits/2159408d4dbbe69b8dffc7c77b3884292f036166

dune-eng commented:
Workflow run id 11001525630 approved.

dune-eng commented:
Workflow run id 11001525861 approved.

dune-eng commented:
Workflow run id 11001960009 approved.

Synthquest commented:
@jeff-dude 
1) removed local sources
2) updated existing sources: https://github.com/duneanalytics/spellbook/pull/6757/commits/2159408d4dbbe69b8dffc7c77b3884292f036166

3) Rechecked from shorter time frame

Query is currently running tests but is set for the full time frame
lmk if there's anything else 🙏
